metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala

```diff
@@ -35,138 +33,152 @@ import scala.meta.internal.semanticdb.Synthetic
+          definitionPath = definitionLocation.getUri().toAbsolutePath
+          if canRenameSymbol(occurence.symbol, None, definitionPath)
+          range <- occurence.range
+        } yield range.toLSP
+      }
     }
   }
 
-  def rename(params: RenameParams): WorkspaceEdit = {
+  def rename(
+      params: RenameParams,
+      token: CancelToken
+  ): Future[WorkspaceEdit] = {
     if (!compilations.currentlyCompiling.isEmpty) {
       client.showMessage(isCompiling)
-      new WorkspaceEdit()
+      Future.successful(new WorkspaceEdit())
     } else {
       val source = params.getTextDocument.getUri.toAbsolutePath
-      val textParams = new TextDocumentPositionParams(
-        params.getTextDocument(),
-        params.getPosition()
-      )
+      definitionProvider.definition(source, params, token).map { definition =>
+        val textParams = new TextDocumentPositionParams(
+          params.getTextDocument(),
+          params.getPosition()
+        )
 
-      val symbolOccurence =
-        definitionProvider.symbolOccurrence(source, textParams.getPosition)
+        val symbolOccurence =
+          definitionProvider.symbolOccurrence(source, textParams.getPosition)
 
-      val suggestedName = params.getNewName()
-      val newName =
-        if (suggestedName.charAt(0) == '`')
-          suggestedName.substring(1, suggestedName.length() - 1)
-        else suggestedName
+        val suggestedName = params.getNewName()
+        val newName =
+          if (suggestedName.charAt(0) == '`')
+            suggestedName.substring(1, suggestedName.length() - 1)
+          else suggestedName
 
-      def includeSynthetic(syn: Synthetic) = {
-        syn.tree match {
-          case SelectTree(_, id) =>
-            id.exists(_.symbol.desc.name.toString == "apply")
-          case _ => false
+        def includeSynthetic(syn: Synthetic) = {
+          syn.tree match {
+            case SelectTree(_, id) =>
+              id.exists(_.symbol.desc.name.toString == "apply")
+            case _ => false
+          }
         }
-      }
 
-      val allReferences = for {
-        (occurence, semanticDb) <- symbolOccurence.toIterable
-        if canRenameSymbol(occurence.symbol, Option(newName))
-        parentSymbols = implementationProvider
-          .topMethodParents(occurence.symbol, semanticDb)
-        txtParams <- {
-          if (parentSymbols.isEmpty) List(textParams)
-          else parentSymbols.map(toTextParams)
-        }
-        isLocal = occurence.symbol.isLocal
-        currentReferences = referenceProvider
-          .references(
-            // we can't get definition by name for local symbols
-            toReferenceParams(txtParams, includeDeclaration = isLocal),
-            canSkipExactMatchCheck = false,
-            includeSynthetics = includeSynthetic
+        val allReferences = for {
+          (occurence, semanticDb) <- symbolOccurence.toIterable
+          definitionLoc <- definition.locations.asScala
```

Comment:
I would not go with `<-` here as it means if you get more locations all references and will be calculated twice for each location

Reply:
I should most likely go with headOption, but we don't have really currently a situation with multiple definitions.